### PR TITLE
Convert sharable types into sendable types (and vice versa)

### PR DESCRIPTION
**Pull request by segeljakt**
For once, I present to you some really dank code

### CODE REVIEW

**Review by frej — APPROVED** [review]
LGTM. Is it limitations in the macro system that prevents you from inspecting the types of the struct fields? At the MLIR-level I don't think we'll have any problems generating the annotations by just looking at the types.

**Comment by frej** [comment]
Oh, the tests failed, I guess we'll have to implement the annotations in the MLIR code generator.

**Comment by segeljakt** [comment]
> LGTM. Is it limitations in the macro system that prevents you from inspecting the types of the struct fields? At the MLIR-level I don't think we'll have any problems generating the annotations by just looking at the types.

It is possible to inspect the types syntactically and map them to their categories. Primitives should probably be hard coded into the code generator. The distinction between algebraic and dynamic types is only about performance. We for example do not need box non-recursive types `Box<String>`. Still, maybe it is a good idea to box them to simplify the generated code.

Currently we have that `impl From<Operable> for Sendable` and `impl From<Sendable> for Operable`. Where

```rust
trait From<T> {
    fn from(x: T) -> Self;
}
```

This means multiple conversions could exist per type. I am going to change this to a new trait `Convert`:

```rust
trait Convert {
    type T;
    fn convert(self) -> Self::T;
}
```

And have `impl Convert for Sendable { type T = Operable }` and `impl Convert for Operable { type T = Sendable }`. This ensures only a single conversion exists per type. Because of this, we can also retrieve the corresponding type after conversion for any type. This however requires that all types, including primitives, implement the trait. I think it can be a good approach however since it gives us full control over which types are allowed and not allowed to run on our system.

**Comment by segeljakt** [comment]
Made some dank mods to the conversion code, now it should be way easier to generate

**Comment by segeljakt** [comment]
We can bring arc-mlir up-to-date next week

**Comment by segeljakt** [comment]
Now we can have some fun and send recursive types over the network OwO'

**Comment by segeljakt** [comment]
I could undo the last commit if you do not like it (added it to make the generated code less verbose).

File: arc-mlir/src/include/Rust/RustPrinterStream.h [hunk 1]
```diff
@@ -139,8 +139,8 @@ class RustPrinterStream {
       if (t.isa<FunctionType>() || t.isa<types::RustEnumType>() ||
           t.isa<types::RustStructType>() || t.isa<types::RustTensorType>() ||
           t.isa<types::RustTupleType>())
-        return "v" + std::to_string(id) + ".clone()";
-      return "v" + std::to_string(id);
+        return "arcorn::val!(v" + std::to_string(id) + ")";
+      return "arcorn::val!(v" + std::to_string(id) + ")";
```
Comment (frej) [inline arc-mlir/src/include/Rust/RustPrinterStream.h]: With `acorn::val!` used for all types, special-casing the non-primitive types is no longer needed, so drop the `if`.

**Review by frej — APPROVED** [review]
LGTM when you fix the inline comment.

> To make the generated code less verbose I removed the arcorn:: prefix from paths

That's an improvement, so go for it.

File: arc-mlir/src/include/Rust/RustPrinterStream.h (hunk 1 above)
Comment (segeljakt) [reply arc-mlir/src/include/Rust/RustPrinterStream.h]: Ok, it's an abstraction with "hopefully no regret" 😅

**Pull request merged**